### DISCUSSION TITLE
Using hostname instead of IP address for broker does not work

### DISCUSSION THREAD

**Discussion by focher** [Resinchem/Tailwind2MQTT#8] [General]
When using a hostname for the `MQTTSERVER` attribute in the `Credentials.h` file, the commands from Home Assistant do not work.

For example, setting `MQTTSERVER` to `mqtt-broker.mydomain` seemingly allows the client to register with the broker. However, sending any commands from Home Assistant does not take affect. Reverting back to using the IP address of the broker instead of hostname restores all functionality.

I will do further testing when I have time, but my internal network is configured with a Pi-hole DNS server where I've added the mqtt-broker hostname to the DNS and I do see the client connect to the broker even when using the broker's hostname. It just seems to be either from HA to the broker or from the broker to the client that something stops working.

**Comment by Resinchem** [comment]
Thanks for the info and additional testing.  I'm a bit perplexed as to what is happening as well.  I also run Pi-Hole on my internal network, but rarely use host names and rely on IP addresses.  I'm curious, when you get the chance to test again, about a couple of things...

First, when using host name, can you see received topics using MQTT explorer from both the HA side and from the bridge/client side?  I have a sneaking suspicion that it has something to do with how MQTT discovery works in HA.  It might mean that you have to remove and recreate the entities in HA after changing the MQTTSERVER settings in Credentials.  It shouldn't happen this way, but I'm thinking that when the entities are auto-created in HA via discovery, broker info is somehow part of that.

Probably the easiest way to test this is to set the ha_discovery flag to FALSE in Settings.h while still using the original IP address for MQTTSERVER.  Then upload/restart the bridge.  This should remove all the auto-created entities in HA (note that this will temporarily break any Lovelace cards/automations).  Once they are removed, change your MQTTSERVER to the host name and set the ha_discovery flag back to TRUE and once again upload/restart.  This should recreate the entities, this time with the host name as the MQTTSERVER.

Again, I really don't know the inter-workings of how HA Discovery works.  But my gut tells me that info about the broker is part of the entity auto-creation process.  So, if you change the broker reference from IP to host, but the originally created entities are still using IP address, it causing that drop in communication somewhere.  Technically, this makes no sense, as you should be able to send topics to the broker using either IP or hostname... but that's my gut instinct.  Again, MQTT Explorer may be of use here.  I'm not sure I will be able to duplicate since I use the MQTT add-on in Home Assistant, and therefore the broker IP address and hostname are the same as my Home Assistant installation.

I'll do a little more poking around when I get the chance, but I think removing the auto-discovered entities and having them recreated AFTER changing the MQTTSERVER setting will resolve the issue.
